src/corppa/poetry_detection/poem_excerpts.py

```diff
@@ -0,0 +1,45 @@
+"""
+Custom data type for poetry excerpts identified with the text of PPA pages.
+"""
+
+from dataclasses import asdict, dataclass
+from typing import Any, Optional
+
+
+@dataclass(kw_only=True)
+class PoemExcerpt:
+    """
+    A poem excerpt data type representing an identified poem excerpt within a
+    PPA page
+    """
+
+    # PPA page related
+    page_id: str
+    ppa_span_start: int
+    ppa_span_end: int
+    ppa_span_text: str
+    # Reference poem related
+    poem_id: str
+    ref_corpus: str
+    ref_span_start: Optional[int] = None
+    ref_span_end: Optional[int] = None
+    ref_span_text: Optional[str] = None
+    # Methods & additional notes
+    detection_methods: set[str]
+    identification_methods: set[str]
+    notes: Optional[str] = None
+
+    def to_dict(self) -> dict[str, Any]:
+        """
+        Returns a JSON-friendly dict of the poem excerpt. Note that unset optional fields
+        are not included.
+        """
+        json_dict = asdict(self)
+        json_dict["detection_methods"] = list(self.detection_methods)
+        json_dict["identification_methods"] = list(self.identification_methods)
+
+        # Remove optional fields if set to None
+        for field in ["ref_span_start", "ref_span_end", "ref_span_text", "notes"]:
+            if json_dict[field] is None:
+                del json_dict[field]
+        return json_dict
```

Comment:
fwiw, I would write it this way, which I think is more pythonic and is agnostic about which fields are optional:

But your version is fine for now!

```suggestion
        return {k:v for k, v in json_dict.items() if v is not None}
```

Reply:
My one worry here is we ever have a required field that can be None... this would remove those ones. Which is fine as long as we explicitly document this behavior. It was quite a surprise when I learned that the `text` field is not required for the ppa page jsonl structure